### PR TITLE
[MetricKit] Add new framework up to Xcode11 GM.

### PR DESCRIPTION
**Pull request by mandel-macaque**

### CODE REVIEW

File: src/metrickit.cs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+using CoreFoundation;
+using Foundation;
+using ObjCRuntime;
+
+namespace MetricKit {
+
+	interface NSUnitDuration : NSUnit { }
+	interface NSUnitInformationStorage : NSUnit { }
+
+	[NoWatch, NoTV, NoMac, iOS (13,0)]
```
Comment (jamesmontemagno) [inline src/metrickit.cs]: Should this have the catalyst annotation on it?

File: src/Constants.iOS.cs.in [hunk 2]
```diff
@@ -143,5 +143,6 @@ namespace MonoTouch {
 		public const string LinkPresentationLibrary = "/System/Library/Frameworks/LinkPresentation.framework/LinkPresentation";
 		public const string VisionKitLibrary = "/System/Library/Frameworks/VisionKit.framework/VisionKit";
 		public const string CoreHapticsLibrary = "/System/Library/Frameworks/CoreHaptics.framework/CoreHaptics";
+		public const string MetricKitLibrary = "/System/Library/Frameworks/MetricKit.framework/MetricKit";
```
Comment (spouliot) [inline src/Constants.iOS.cs.in]: please keep them sorted to ease comparison with the other files

File: src/metrickit.cs [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+	[BaseType (typeof(NSObject))]
+	interface MXHistogram<UnitType> : NSSecureCoding
+		where UnitType : NSUnit {
+		[Export ("totalBucketCount")]
+		nuint TotalBucketCount { get; }
+
+		[Export ("bucketEnumerator", ArgumentSemantic.Strong)]
+		NSEnumerator<MXHistogramBucket<NSUnit>> BucketEnumerator { get; }
```
Comment (spouliot) [inline src/metrickit.cs]: `UnitType` not `NSUnit`
```objc
NSEnumerator<MXHistogramBucket<UnitType> *>
```

File: src/metrickit.cs [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (13,0)]
+	[BaseType (typeof(NSObject))]
+	interface MXMetaData : NSSecureCoding {
+		[Export ("regionFormat", ArgumentSemantic.Strong)]
+		string RegionFormat { get; }
+
+		[Export ("osVersion", ArgumentSemantic.Strong)]
+		string OsVersion { get; }
```
Comment (spouliot) [inline src/metrickit.cs]: `OSVersion` -> acronym with only two letters stays capitalized in .net

File: src/metrickit.cs [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+	[NoWatch, NoTV, NoMac, iOS (13,0)]
+	[BaseType (typeof(NSObject))]
+	interface MXMetricManager {
+		[Export ("pastPayloads", ArgumentSemantic.Strong)]
+		MXMetricPayload[] PastPayloads { get; }
+
+		[Static]
+		[Export ("sharedManager", ArgumentSemantic.Strong)]
+		MXMetricManager SharedManager { get; }
```
Comment (spouliot) [inline src/metrickit.cs]: ^ that hints you should not create your own -> `[DisableDefaultCtor]`

File: src/metrickit.cs [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+	[BaseType (typeof(MXMetric), Name = "MXGPUMetric")]
+	interface MXGpuMetric {
+		[Export ("cumulativeGPUTime", ArgumentSemantic.Strong)]
+		NSMeasurement<NSUnitDuration> CumulativeGpuTime { get; }
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (13,0)]
+	[BaseType (typeof(NSDimension))]
+	[DisableDefaultCtor]
+	interface MXUnitSignalBars : NSUnit {
```
Comment (spouliot) [inline src/metrickit.cs]: why `: NSUnit` ? the base type is set with an attribute and `NSUnit` is not a protocol

File: src/metrickit.cs [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+
+		[Static]
+		[Export ("bars", ArgumentSemantic.Copy)]
+		MXUnitSignalBars Bars { get; }
+	}
+
+	[NoWatch, NoTV, NoMac, iOS (13,0)]
+	[BaseType (typeof(NSDimension))]
+	[DisableDefaultCtor]
+	interface MXUnitAveragePixelLuminance : NSUnit {
```
Comment (spouliot) [inline src/metrickit.cs]: same

File: src/metrickit.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/metrickit.cs]: Headers don't say it, but docs do.. :/

File: src/metrickit.cs (hunk 6 above)
Comment (mandel-macaque) [reply src/metrickit.cs]: If I do so, I get the following compilation error:

`
 The type 'MetricKit.MXUnitSignalBars' cannot be used as type parameter 'UnitType' in the generic type or method 'MXHistogram<UnitType>'. There is no implicit reference conversion from 'MetricKit.MXUnitSignalBars' to 'Foundation.NSUnit'.
`

Comment (mandel-macaque) [reply src/metrickit.cs]: The generated code with my code looks good: 

```csharp
[Register("MXUnitSignalBars", true)]
        [Unavailable (PlatformName.WatchOS, PlatformArchitecture.All)]
        [Unavailable (PlatformName.TvOS, PlatformArchitecture.All)]
        [Unavailable (PlatformName.MacOSX, PlatformArchitecture.All)]
        [Introduced (PlatformName.iOS, 13,0, PlatformArchitecture.All)]
        public unsafe partial class MXUnitSignalBars : NSDimension, INSCoding, INSCopying, INSSecureCoding {
                
                [BindingImpl (BindingImplOptions.GeneratedCode | BindingImplOptions.Optimizable)]
                static readonly IntPtr class_ptr = Class.GetHandle ("MXUnitSignalBars");
                
                public override IntPtr ClassHandle { get { return class_ptr; } }
                
                [BindingImpl (BindingImplOptions.GeneratedCode | BindingImplOptions.Optimizable)]
```

Comment (dalexsoto) [reply src/metrickit.cs]: I think this is fine, what is happening is that temp.dll build does no know about inheritance chain yet due to at this point is metadata, by doing `interface MXUnitSignalBars : NSUnit` you create that link but the generator will do the right thing and ignore it. I would still add a comment to the code tho as @spouliot correctly points out `NSUnit` is not a protocol so better remind our future selves why we did this.

File: src/Constants.iOS.cs.in (hunk 2 above)
Comment (dalexsoto) [reply src/Constants.iOS.cs.in]: Not much of a sort order here heh, I guess @mandel-macaque  can do the hardwork sort here :P

File: src/metrickit.cs (hunk 1 above)
Comment (dalexsoto) [reply src/metrickit.cs]: Yeah we tend to follow headers here since it is what is is actually shipped and used.

Comment (spouliot) [reply src/metrickit.cs]: we'll need more tooling (both intro and xtro at least) to cover what's in catalyst and what's not - docs and headers are never enough :(

**Pull request merged**